sites/cheerpj/src/content/docs/13-tutorials/05-interoperability-tutorial.mdx

```diff
@@ -0,0 +1,159 @@
+
+## Project Structure
+
+The project consists of the following files and directories:
+
+```plaintext
+CheerpJInteroperbailityTutorial/
+├── com/example
+│   └── Example.java
+│   └── manifest.txt
+├── index.html
+└── example.jar
+└── Makefile
+```
+
+## Setting Up the Project
+
+We start by creating a simple Java class with a native method that will interact with JavaScript. This class will receive input from JavaScript, process it, and send a response back to JavaScript.
+
+```java title="Example.java"
+package com.example;
+
+public class Example {
+    public static native void sendToHTML(String s);
+
+    public String processInput(String input) {
+        sendToHTML(input);
+        return "Java received: " + input;
+    }
+}
+```
+
+In the `Example` class, we define a native method `sendToHTML` that will be implemented in JavaScript. The `processInput` method receives input from JavaScript, sends it to the `sendToHTML` method, and returns a response.
+
+## Implementing the JavaScript Side
+
+Next, we create a JavaScript function that receives input from Java, displays it in the HTML document, and logs the input to the console.
+
+```js title="index.html"
+async function Java_com_example_Example_sendToHTML(lib, str) {
+	document.getElementById("javaOutput").innerText =
+		"JavaScript received: " + str;
+	console.log("Received input from Java: " + str);
+}
+```
+
+Here, the `Java_com_example_Example_sendToHTML` function in JavaScript acts as a native method, allowing a Java class to communicate with JavaScript.
+
+## Handling Multiple JavaScript Instances
+
+To establish persistent communication between Java and JavaScript, we’ll set up a long-running Java thread that enables JavaScript to access Java methods continuously. For example, using `cheerpjRunLibrary` to execute a Java class with native methods creates a separate process from `cheerpjRunJar`, which can prevent resources from being shared between instances.
+
+### Setting the Java Application Instance in JavaScript
+
+To establish consistent communication between Java and JavaScript instances, we use a long-running Java thread within the Java application.
+
+```java title="Example.java"
+public class Example {
+  public static native void nativeSetApplication(Example myApplication);
+
+  public static void main(String[] args) {
+        Example app = new Example();
+        new Thread(() -> {
+            nativeSetApplication(app);
+            System.out.println("Starting Thread");
+        }).start();
+    }
+}
+```
+
+We also use a native function to capture this thread on the JavaScript side, enabling persistent access to Java methods from JavaScript.
+
+```js title="index.html"
+async function Java_com_example_Example_nativeSetApplication(
+	lib,
+	myApplication
+) {
+	window.myApplication = myApplication;
+	console.log("Java application instance set on JavaScript side.");
+	return new Promise(() => {}); // Keeps the function from returning
+}
+```
```

Comment:
I think it's important here to show a bigger part of the HTML file. We could also point out that you could make a separate `.js` file, and also add that as a script if people want to keep the HTML file clean.